File: flutter_mobile/lib/utils/vc_slot_window.dart
/// Client-side helpers for video-consult join window (mirrors server IST logic).
class VcSlotWindow {
  const VcSlotWindow({
    this.slotStartAt,
    this.slotEndAt,
    this.graceEndsAt,
    this.joinOpensAt,
    this.canJoinWindow = true,
    this.inGrace = false,
    this.softWarn = false,
    this.forceEnd = false,
    this.windowMessage,
  });

  final DateTime? slotStartAt;
  final DateTime? slotEndAt;
  final DateTime? graceEndsAt;
  final DateTime? joinOpensAt;
  final bool canJoinWindow;
  final bool inGrace;
  final bool softWarn;
  final bool forceEnd;
  final String? windowMessage;

  static const slotMinutes = 15;
  static const earlyGraceMinutes = 1;
  static const lateGraceMinutes = 5;
  static const softWarnBeforeEndMinutes = 2;

  factory VcSlotWindow.fromJson(Map<String, dynamic>? json) {
    if (json == null) return const VcSlotWindow();
    DateTime? ms(dynamic v) {
      if (v == null) return null;
      final n = v is num ? v.toInt() : int.tryParse('$v');
      if (n == null || n <= 0) return null;
      return DateTime.fromMillisecondsSinceEpoch(n);
    }

    return VcSlotWindow(
      slotStartAt: ms(json['slotStartAt'] ?? json['slot_start_at']),
      slotEndAt: ms(json['slotEndAt'] ?? json['slot_end_at']),
      graceEndsAt: ms(json['graceEndsAt'] ?? json['grace_ends_at']),
      joinOpensAt: ms(json['joinOpensAt'] ?? json['join_opens_at']),
      canJoinWindow: json['canJoinWindow'] != false,
      inGrace: json['inGrace'] == true,
      softWarn: json['softWarn'] == true,
      forceEnd: json['forceEnd'] == true,
      windowMessage: json['windowMessage']?.toString() ?? json['window_message']?.toString(),
    );
  }

  /// Compute from appointment slot_date / slot_time when server window is absent.
  factory VcSlotWindow.fromSlot({
    required String slotDate,
    required String slotTime,
    DateTime? now,
  }) {
    final bounds = parseSlotBounds(slotDate, slotTime);
    if (bounds == null) return const VcSlotWindow();
    final start = bounds.$1;
    final end = bounds.$2;
    final n = now ?? DateTime.now();
    final joinOpens = start.subtract(const Duration(minutes: earlyGraceMinutes));
    final graceEnds = end.add(const Duration(minutes: lateGraceMinutes));
    final softAt = end.subtract(const Duration(minutes: softWarnBeforeEndMinutes));
    final force = !n.isBefore(graceEnds);
    final grace = !n.isBefore(end) && n.isBefore(graceEnds);
    final soft = !force && !n.isBefore(softAt);
    final canJoin = !n.isBefore(joinOpens) && n.isBefore(graceEnds);

    String? message;
    if (n.isBefore(joinOpens)) {
      message = 'Join opens at ${_fmtTime(start)}';
    } else if (force) {
      message = 'This slot has ended';
    } else if (grace) {
      message = 'Slot ended. Grace period — finish or leave soon.';
    } else if (soft) {
      message = 'Consultation ends in 2 minutes.';
    }

    return VcSlotWindow(
      slotStartAt: start,
      slotEndAt: end,
      graceEndsAt: graceEnds,
      joinOpensAt: joinOpens,
      canJoinWindow: canJoin,
      inGrace: grace,
      softWarn: soft,
      forceEnd: force,
      windowMessage: message,
    );
  }

  static String _fmtTime(DateTime dt) {
    final h = dt.hour % 12 == 0 ? 12 : dt.hour % 12;
    final m = dt.minute.toString().padLeft(2, '0');
    final mer = dt.hour >= 12 ? 'PM' : 'AM';
    return '$h:$m $mer';
  }

  static (DateTime, DateTime)? parseSlotBounds(String slotDate, String slotTime) {
    final day = _parseLegacyDate(slotDate);
    if (day == null) return null;
    final raw = slotTime.trim();
    if (raw.isEmpty) return null;

    final parts = raw.split(RegExp(r'\s*[-–—]\s*'));
    final startClock = _parseClock(parts.first);
    if (startClock == null) return null;
    final start = DateTime(day.year, day.month, day.day, startClock.$1, startClock.$2);

    DateTime? end;
    if (parts.length > 1) {
      final endClock = _parseClock(parts[1]);
      if (endClock != null) {
        end = DateTime(day.year, day.month, day.day, endClock.$1, endClock.$2);
        if (!end.isAfter(start)) {
          end = end.add(const Duration(days: 1));
        }
      }
    }
    end ??= start.add(const Duration(minutes: slotMinutes));
    return (start, end);
  }

  static DateTime? _parseLegacyDate(String slotDate) {
    final raw = slotDate.trim().replaceAll('/', '_').replaceAll('-', '_');
    final parts = raw.split('_');
    if (parts.length != 3) return null;
    final d = int.tryParse(parts[0]);
    final m = int.tryParse(parts[1]);
    final y = int.tryParse(parts[2]);
    if (d == null || m == null || y == null) return null;
    return DateTime(y, m, d);
  }

  static (int, int)? _parseClock(String text) {
    final s = text.trim();
    if (s.isEmpty) return null;
    final m = RegExp(r'(\d{1,2})\s*:\s*(\d{2})\s*(AM|PM|am|pm)?').firstMatch(s);
    if (m != null) {
      var hour = int.parse(m.group(1)!);
      final minute = int.parse(m.group(2)!);
      final mer = (m.group(3) ?? '').toUpperCase();
      if (mer == 'PM' && hour < 12) hour += 12;
      if (mer == 'AM' && hour == 12) hour = 0;
      return (hour, minute);
    }
    final m2 = RegExp(r'(\d{1,2})\s*(AM|PM|am|pm)').firstMatch(s);
    if (m2 != null) {
      var hour = int.parse(m2.group(1)!);
      final mer = m2.group(2)!.toUpperCase();
      if (mer == 'PM' && hour < 12) hour += 12;
      if (mer == 'AM' && hour == 12) hour = 0;
      return (hour, 0);
    }
    return null;
  }

  String get joinButtonLabel {
    if (forceEnd) return 'Slot ended';
    if (!canJoinWindow && joinOpensAt != null) {
      final mins = joinOpensAt!.difference(DateTime.now()).inMinutes;
      if (mins > 0) return 'Opens in ${mins}m';
      return 'Join opens soon';
    }
    return 'Join Video Call';
  }

  String leaveConfirmMessage() {
    final end = slotEndAt;
    if (end == null) {
      return 'Leave the call? You can rejoin until the slot ends.';
    }
    return 'Leave the call? You can rejoin until ${_fmtTime(end)}.';
  }
}

File: fastapi_back/app/utils/vc_slot_window.py
"""Video consult join window helpers (IST wall-clock).

Join allowed in [slot_start - 1 min, slot_end + 5 min].
Soft warning at T−2 min before slot_end; force-end at grace end.
"""
from __future__ import annotations

import re
from datetime import datetime, timedelta
from typing import Any, Optional, Tuple
from zoneinfo import ZoneInfo

IST = ZoneInfo("Asia/Kolkata")

VC_SLOT_MINUTES = 15
EARLY_GRACE_MINUTES = 1
LATE_GRACE_MINUTES = 5
SOFT_WARN_BEFORE_END_MINUTES = 2


def now_ist_naive() -> datetime:
    return datetime.now(IST).replace(tzinfo=None)


def _parse_legacy_date(slot_date: str) -> Optional[datetime]:
    raw = str(slot_date or "").strip().replace("/", "_").replace("-", "_")
    parts = raw.split("_")
    if len(parts) != 3:
        return None
    try:
        d, m, y = int(parts[0]), int(parts[1]), int(parts[2])
        return datetime(y, m, d)
    except ValueError:
        return None


def _parse_clock(text: str) -> Optional[Tuple[int, int]]:
    """Parse a time fragment into (hour, minute) 24h."""
    s = (text or "").strip()
    if not s:
        return None
    m = re.search(r"(\d{1,2})\s*:\s*(\d{2})\s*(AM|PM|am|pm)?", s)
    if m:
        hour = int(m.group(1))
        minute = int(m.group(2))
        mer = (m.group(3) or "").upper()
        if mer == "PM" and hour < 12:
            hour += 12
        if mer == "AM" and hour == 12:
            hour = 0
        if 0 <= hour <= 23 and 0 <= minute <= 59:
            return hour, minute
    m2 = re.search(r"(\d{1,2})\s*(AM|PM|am|pm)", s)
    if m2:
        hour = int(m2.group(1))
        mer = m2.group(2).upper()
        if mer == "PM" and hour < 12:
            hour += 12
        if mer == "AM" and hour == 12:
            hour = 0
        if 0 <= hour <= 23:
            return hour, 0
    nums = [int(x) for x in re.findall(r"\d+", s)]
    if nums:
        hour = nums[0]
        minute = nums[1] if len(nums) > 1 else 0
        lower = s.lower()
        if "pm" in lower and hour < 12:
            hour += 12
        if "am" in lower and hour == 12:
            hour = 0
        if 0 <= hour <= 23 and 0 <= minute <= 59:
            return hour, minute
    return None


def parse_slot_bounds(slot_date: str, slot_time: str) -> Optional[Tuple[datetime, datetime]]:
    """Return (slot_start, slot_end) naive IST wall clock, or None."""
    day = _parse_legacy_date(slot_date)
    if day is None:
        return None
    raw = str(slot_time or "").strip()
    if not raw:
        return None

    # Range: "2:00 PM - 2:15 PM" / "14:00 - 14:15"
    parts = re.split(r"\s*[-–—]\s*", raw, maxsplit=1)
    start_clock = _parse_clock(parts[0])
    if start_clock is None:
        return None
    start = day.replace(hour=start_clock[0], minute=start_clock[1], second=0, microsecond=0)

    end: Optional[datetime] = None
    if len(parts) > 1:
        end_clock = _parse_clock(parts[1])
        if end_clock is not None:
            end = day.replace(hour=end_clock[0], minute=end_clock[1], second=0, microsecond=0)
            if end <= start:
                end = end + timedelta(days=1)

    if end is None:
        end = start + timedelta(minutes=VC_SLOT_MINUTES)
    return start, end


def slot_window_payload(
    appointment: dict,
    *,
    now: Optional[datetime] = None,
) -> dict[str, Any]:
    """Compute join/soft-end/force-end flags for an appointment."""
    now = now or now_ist_naive()
    bounds = parse_slot_bounds(
        str(appointment.get("slot_date") or ""),
        str(appointment.get("slot_time") or ""),
    )
    if bounds is None:
        # Fail open for malformed legacy rows so existing bookings still work.
        return {
            "slotStartAt": None,
            "slotEndAt": None,
            "graceEndsAt": None,
            "joinOpensAt": None,
            "canJoinWindow": True,
            "inGrace": False,
            "softWarn": False,
            "forceEnd": False,
            "windowMessage": None,
            "slotDurationMinutes": VC_SLOT_MINUTES,
            "earlyGraceMinutes": EARLY_GRACE_MINUTES,
            "lateGraceMinutes": LATE_GRACE_MINUTES,
        }

    slot_start, slot_end = bounds
    join_opens = slot_start - timedelta(minutes=EARLY_GRACE_MINUTES)
    grace_ends = slot_end + timedelta(minutes=LATE_GRACE_MINUTES)
    soft_at = slot_end - timedelta(minutes=SOFT_WARN_BEFORE_END_MINUTES)

    force_end = now >= grace_ends
    in_grace = slot_end <= now < grace_ends
    soft_warn = (not force_end) and now >= soft_at
    can_join = join_opens <= now < grace_ends

    message = None
    if now < join_opens:
        message = f"Join opens at {slot_start.strftime('%I:%M %p').lstrip('0')}"
    elif force_end:
        message = "This slot has ended"
    elif in_grace:
        message = "Slot ended. Grace period — finish or leave soon."
    elif soft_warn:
        message = "Consultation ends in 2 minutes."

    def _ms(dt: datetime) -> int:
        # Clients treat these as absolute epoch ms in local IST wall clock.
        aware = dt.replace(tzinfo=IST)
        return int(aware.timestamp() * 1000)

    return {
        "slotStartAt": _ms(slot_start),
        "slotEndAt": _ms(slot_end),
        "graceEndsAt": _ms(grace_ends),
        "joinOpensAt": _ms(join_opens),
        "canJoinWindow": can_join,
        "inGrace": in_grace,
        "softWarn": soft_warn,
        "forceEnd": force_end,
        "windowMessage": message,
        "slotDurationMinutes": max(
            1, int((slot_end - slot_start).total_seconds() // 60)
        ),
        "earlyGraceMinutes": EARLY_GRACE_MINUTES,
        "lateGraceMinutes": LATE_GRACE_MINUTES,
    }


def assert_within_join_window(appointment: dict) -> Optional[str]:
    """Return an error message if outside the VC join window."""
    window = slot_window_payload(appointment)
    if window.get("canJoinWindow"):
        return None
    return window.get("windowMessage") or "Video call is not available for this slot right now"
